ocviapy/__init__.py

```diff
@@ -595,58 +611,78 @@ def __init__(self, namespace, restype, name, watch_owned=False, watcher=None):
-
-                if not previously_observed and not resource.ready:
-                    log.info(
-                        "[%s] found owned resource %s, not yet ready",
-                        self.key,
-                        resource.key,
-                    )
-
-                # check if ready state has transitioned for this resource
-                if not previously_ready and resource.ready:
-                    log.info("[%s] owned resource %s is ready!", self.key, resource.key)
+                return True
+        return False
+
+    def _check_status_if_owned(self, resource):
+        if self._owns_resource(resource):
+            previously_observed = False
+            previously_ready = False
+            if resource.key in self.observed_resources:
+                # so we don't keep logging on every loop
+                previously_observed = True
+                previous_resource = self.observed_resources[resource.key]
+                previously_ready = True if previous_resource.ready else False
+
+            # update our records for this resource
+            self.observed_resources[resource.key] = resource
+
+            if not previously_observed and not resource.ready:
+                log.info(
+                    "[%s] found owned resource %s, not yet ready",
+                    self.key,
+                    resource.key,
+                )
+
+            # check if ready state has transitioned for this resource
+            if not previously_ready and resource.ready:
+                log.info("[%s] owned resource %s is ready!", self.key, resource.key)
 
     @property
     def _all_resources_ready(self):
         resources_ready = [r.ready is True for _, r in self.observed_resources.items()]
         return len(resources_ready) > 0 and all(resources_ready)
 
+    def _check_for_status_errors(self, resource):
+        error = resource.image_pull_error
+        if error:
+            self.status_errors.add(error)
+
+    def _check_owned_resources(self):
+        # use .copy() in case dict changes during iteration
+        for _, r in self.watcher.resources.copy().items():
+            self._check_for_status_errors(r)
+            self._check_status_if_owned(r)
+
+        # check to see if any of the owned resources we were previously watching are now no
+        # longer present in the ResourceWatcher
+        disappeared_resources = {
+            key for key in self.observed_resources if key not in self.watcher.resources
+        }
+        for key in disappeared_resources:
+            log.info("[%s] resource has disappeared, no longer monitoring it", key)
+            del self.observed_resources[key]
+
     def _observe(self, resource):
         key = resource.key
-
         # update our records for this resource
         self.observed_resources[key] = resource
 
+        self._check_for_status_errors(resource)
         if self.watch_owned:
-            # use .copy() in case dict changes during iteration
```

Comment:
This code block was moved into the `_check_owned_resources` func. The only new addition to this block of code is https://github.com/RedHatInsights/ocviapy/pull/9/files#diff-028ea95c402cb62a8d41c1f73660fbfbe0fc44359b8a960ffacb4d54a0abaf2fR666